File: referencia/flutter-app/test/nucleo/calificador_corte_test.dart
// Tests de la calificación según la norma INEN 176.
//
// Los casos vienen de `casos_norma.json`, EL MISMO archivo que usan los tests
// de Python en `entrenamiento/pruebas/test_calificar_corte.py`. Si la lógica de
// Dart se desvía de la de Python, estos tests fallan. Esa es toda la idea: el
// resultado que ve el usuario en el teléfono tiene que ser idéntico al que se
// valida en la computadora contra los tableros contados por un experto.

import 'dart:convert';
import 'dart:io';

import 'package:cacaotrace/nucleo/norma/calificador_corte.dart';
import 'package:cacaotrace/nucleo/norma/tabla_norma.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  late CalificadorCorte calificador;
  late Map<String, dynamic> casos;

  setUpAll(() {
    final tabla = TablaNorma.desdeJsonTexto(
      File('assets/norma/norma_inen176.json').readAsStringSync(),
    );
    calificador = CalificadorCorte(tabla);
    casos = jsonDecode(
      File('test/nucleo/casos_norma.json').readAsStringSync(),
    ) as Map<String, dynamic>;
  });

  group('casos compartidos con la implementación Python', () {
    test('hay casos cargados', () {
      expect((casos['casos'] as List), isNotEmpty);
      expect((casos['errores'] as List), isNotEmpty);
    });

    test('cada caso da el resultado esperado en cada perfil', () {
      for (final caso in casos['casos'] as List) {
        final c = caso as Map<String, dynamic>;
        final conteo = (c['conteo'] as Map).map(
          (k, v) => MapEntry(k as String, v as int),
        );
        (c['espera'] as Map).forEach((perfil, esperado) {
          final e = esperado as Map<String, dynamic>;
          final r = calificador.calificar(conteo, perfil: perfil as String);
          expect(
            r.resultado,
            e['resultado'],
            reason: '${c['nombre']} / $perfil',
          );
          expect(r.conforme, e['conforme'], reason: '${c['nombre']} / $perfil');
        });
      }
    });

    test('los porcentajes coinciden con los esperados', () {
      for (final caso in casos['casos'] as List) {
        final c = caso as Map<String, dynamic>;
        final conteo = (c['conteo'] as Map).map(
          (k, v) => MapEntry(k as String, v as int),
        );
        final pct = calificador.porcentajes(conteo);
        (c['porcentajes'] as Map).forEach((indicador, valor) {
          expect(
            pct[indicador as String],
            closeTo((valor as num).toDouble(), 0.05),
            reason: '${c['nombre']}: $indicador',
          );
        });
        expect(
          pct.granosEvaluados,
          c['granos_evaluados'],
          reason: '${c['nombre']}: granos evaluados',
        );
        if (c.containsKey('granos_ignorados')) {
          expect(pct.granosIgnorados, c['granos_ignorados']);
        }
      }
    });

    test('los avisos coinciden en cantidad', () {
      for (final caso in casos['casos'] as List) {
        final c = caso as Map<String, dynamic>;
        if (!c.containsKey('avisos_esperados')) continue;
        final conteo = (c['conteo'] as Map).map(
          (k, v) => MapEntry(k as String, v as int),
        );
        final perfil = (c['espera'] as Map).keys.first as String;
        final r = calificador.calificar(conteo, perfil: perfil);
        expect(
          r.avisos.length,
          c['avisos_esperados'],
          reason: '${c['nombre']}: ${r.avisos}',
        );
      }
    });

    test('las entradas inválidas lanzan ErrorNorma', () {
      for (final caso in casos['errores'] as List) {
        final c = caso as Map<String, dynamic>;
        final conteo = (c['conteo'] as Map).map(
          (k, v) => MapEntry(k as String, v as int),
        );
        expect(
          () => calificador.calificar(conteo),
          throwsA(isA<ErrorNorma>()),
          reason: c['nombre'] as String,
        );
      }
    });
  });

  group('reglas del cálculo', () {
    test('fermentado_total es la suma de bueno y ligero', () {
      final pct = calificador.porcentajes({
        'bien_fermentado': 50,
        'ligeramente_fermentado': 20,
        'violeta': 30,
      });
      expect(
        pct.fermentadoTotal,
        closeTo(pct['fermentado_bueno'] + pct['fermentado_ligero'], 1e-9),
      );
    });

    test('un perfil inexistente lanza ErrorNorma', () {
      expect(
        () => calificador.calificar({'bien_fermentado': 100},
            perfil: 'no_existe'),
        throwsA(isA<ErrorNorma>()),
      );
    });

    test('una clase desconocida avisa y no rompe el cálculo', () {
      final r = calificador.calificar({
        'bien_fermentado': 90,
        'clase_inventada': 10,
      });
      expect(r.porcentajes.granosEvaluados, 90);
      expect(r.avisos.any((a) => a.contains('clase_inventada')), isTrue);
    });

    test('todos los perfiles de la tabla se pueden calificar', () {
      final conteo = {
        'bien_fermentado': 70,
        'ligeramente_fermentado': 8,
        'violeta': 12,
        'pizarroso': 8,
        'mohoso': 2,
      };
      final todos = calificador.calificarTodos(conteo);
      expect(todos.length, calificador.tabla.perfiles.length);
      for (final r in todos.values) {
        expect(r.resultado, isNotEmpty);
      }
    });

    test('la tabla no usa indicadores que el cálculo no produce', () {
      const validos = {
        'fermentado_bueno',
        'fermentado_ligero',
        'fermentado_total',
        'violeta',
        'pizarroso',
        'mohoso',
        'defectuoso',
      };
      for (final p in calificador.tabla.perfiles.values) {
        final requisitos = p.esPorGrados
            ? p.grados.expand((g) => g.requisitos)
            : p.requisitos;
        for (final r in requisitos) {
          expect(validos, contains(r.indicador),
              reason: 'El perfil ${p.clave} usa "${r.indicador}"');
          expect(['min', 'max'], contains(r.tipo));
        }
      }
    });

    test('la tabla se puede exportar y volver a leer sin perder nada', () {
      final texto = calificador.tabla.aJsonTexto();
      final vuelta = TablaNorma.desdeJsonTexto(texto);
      final c2 = CalificadorCorte(vuelta);
      const conteo = {'bien_fermentado': 70, 'violeta': 20, 'pizarroso': 10};
      expect(
        c2.calificar(conteo).resultado,
        calificador.calificar(conteo).resultado,
      );
    });

    test('avisa cuando se cuentan menos de 97 o más de 103 granos', () {
      final pocos = calificador.calificar({'bien_fermentado': 80});
      expect(pocos.avisos.any((a) => a.contains('80 granos')), isTrue);

      final muchos = calificador.calificar({'bien_fermentado': 110});
      expect(muchos.avisos.any((a) => a.contains('110 granos')), isTrue);

      final justos = calificador.calificar({
        'bien_fermentado': 80,
        'violeta': 20,
      });
      expect(justos.avisos, isEmpty);
    });
  });
}

File: android/nucleo/src/test/kotlin/ec/cacaotrace/nucleo/norma/CalificadorCorteTest.kt
package ec.cacaotrace.nucleo.norma

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestFactory
import org.junit.jupiter.api.assertThrows

/**
 * Tests de la calificación según la norma INEN 176.
 *
 * Los casos vienen de `casos_norma.json`, EL MISMO archivo que usan los tests
 * de Python en `entrenamiento/pruebas/test_calificar_corte.py`. Si la lógica de
 * Kotlin se desvía de la de Python, estos tests fallan. Esa es toda la idea: el
 * resultado que ve el usuario en el teléfono tiene que ser idéntico al que se
 * valida en la computadora contra tableros contados por un experto.
 */
class CalificadorCorteTest {

    private val tabla = TablaNorma.desdeJsonTexto(recurso("norma_inen176.json"))
    private val calificador = CalificadorCorte(tabla)
    private val casos = Json.parseToJsonElement(recurso("casos_norma.json")).jsonObject

    private fun recurso(nombre: String): String =
        checkNotNull(javaClass.classLoader.getResourceAsStream(nombre)) {
            "Falta el recurso de prueba $nombre"
        }.bufferedReader().readText()

    private fun conteoDe(caso: JsonObject): Map<String, Int> =
        caso["conteo"]!!.jsonObject.mapValues { it.value.jsonPrimitive.content.toInt() }

    private fun nombreDe(caso: JsonObject): String = caso["nombre"]!!.jsonPrimitive.content

    @Test
    fun `hay casos cargados`() {
        assertTrue(casos["casos"]!!.jsonArray.isNotEmpty())
        assertTrue(casos["errores"]!!.jsonArray.isNotEmpty())
    }

    @TestFactory
    fun `cada caso da el resultado esperado en cada perfil`(): List<DynamicTest> =
        casos["casos"]!!.jsonArray.map { elemento ->
            val caso = elemento.jsonObject
            DynamicTest.dynamicTest(nombreDe(caso)) {
                val conteo = conteoDe(caso)
                caso["espera"]!!.jsonObject.forEach { (perfil, esperadoJson) ->
                    val esperado = esperadoJson.jsonObject
                    val r = calificador.calificar(conteo, perfil)
                    assertEquals(
                        esperado["resultado"]!!.jsonPrimitive.content,
                        r.resultado,
                        "${nombreDe(caso)} / $perfil",
                    )
                    assertEquals(
                        esperado["conforme"]!!.jsonPrimitive.content.toBoolean(),
                        r.conforme,
                        "${nombreDe(caso)} / $perfil",
                    )
                }
            }
        }

    @TestFactory
    fun `los porcentajes coinciden con los esperados`(): List<DynamicTest> =
        casos["casos"]!!.jsonArray.map { elemento ->
            val caso = elemento.jsonObject
            DynamicTest.dynamicTest(nombreDe(caso)) {
                val pct = calificador.porcentajes(conteoDe(caso))
                caso["porcentajes"]!!.jsonObject.forEach { (indicador, valor) ->
                    assertEquals(
                        valor.jsonPrimitive.content.toDouble(),
                        pct[indicador],
                        0.05,
                        "${nombreDe(caso)}: $indicador",
                    )
                }
                assertEquals(
                    caso["granos_evaluados"]!!.jsonPrimitive.content.toInt(),
                    pct.granosEvaluados,
                    "${nombreDe(caso)}: granos evaluados",
                )
                caso["granos_ignorados"]?.let {
                    assertEquals(it.jsonPrimitive.content.toInt(), pct.granosIgnorados)
                }
            }
        }

    @TestFactory
    fun `los avisos coinciden en cantidad`(): List<DynamicTest> =
        casos["casos"]!!.jsonArray
            .map { it.jsonObject }
            .filter { it.containsKey("avisos_esperados") }
            .map { caso ->
                DynamicTest.dynamicTest(nombreDe(caso)) {
                    val perfil = caso["espera"]!!.jsonObject.keys.first()
                    val r = calificador.calificar(conteoDe(caso), perfil)
                    assertEquals(
                        caso["avisos_esperados"]!!.jsonPrimitive.content.toInt(),
                        r.avisos.size,
                        "${nombreDe(caso)}: ${r.avisos}",
                    )
                }
            }

    @TestFactory
    fun `las entradas invalidas lanzan ErrorNorma`(): List<DynamicTest> =
        casos["errores"]!!.jsonArray.map { elemento ->
            val caso = elemento.jsonObject
            DynamicTest.dynamicTest(nombreDe(caso)) {
                assertThrows<ErrorNorma> { calificador.calificar(conteoDe(caso)) }
            }
        }

    @Test
    fun `fermentado total es la suma de bueno y ligero`() {
        val pct = calificador.porcentajes(
            mapOf(
                "bien_fermentado" to 50,
                "ligeramente_fermentado" to 20,
                "violeta" to 30,
            ),
        )
        assertEquals(
            pct["fermentado_bueno"] + pct["fermentado_ligero"],
            pct.fermentadoTotal,
            1e-9,
        )
    }

    @Test
    fun `un perfil inexistente lanza ErrorNorma`() {
        assertThrows<ErrorNorma> {
            calificador.calificar(mapOf("bien_fermentado" to 100), "no_existe")
        }
    }

    @Test
    fun `una clase desconocida avisa y no rompe el calculo`() {
        val r = calificador.calificar(
            mapOf("bien_fermentado" to 90, "clase_inventada" to 10),
        )
        assertEquals(90, r.porcentajes.granosEvaluados)
        assertTrue(r.avisos.any { it.contains("clase_inventada") })
    }

    @Test
    fun `todos los perfiles de la tabla se pueden calificar`() {
        val conteo = mapOf(
            "bien_fermentado" to 70,
            "ligeramente_fermentado" to 8,
            "violeta" to 12,
            "pizarroso" to 8,
            "mohoso" to 2,
        )
        val todos = calificador.calificarTodos(conteo)
        assertEquals(tabla.perfiles.size, todos.size)
        todos.values.forEach { assertTrue(it.resultado.isNotEmpty()) }
    }

    @Test
    fun `la tabla no usa indicadores que el calculo no produce`() {
        val validos = setOf(
            "fermentado_bueno", "fermentado_ligero", "fermentado_total",
            "violeta", "pizarroso", "mohoso", "defectuoso",
        )
        tabla.perfiles.values.forEach { p ->
            val requisitos =
                if (p.esPorGrados) p.grados.flatMap { it.requisitos } else p.requisitos
            requisitos.forEach { r ->
                assertTrue(
                    r.indicador in validos,
                    "El perfil ${p.clave} usa \"${r.indicador}\"",
                )
                assertTrue(r.tipo in setOf("min", "max"))
            }
        }
    }

    @Test
    fun `la tabla se puede exportar y volver a leer sin perder nada`() {
        val vuelta = TablaNorma.desdeJsonTexto(tabla.aJsonTexto())
        val c2 = CalificadorCorte(vuelta)
        val conteo = mapOf("bien_fermentado" to 70, "violeta" to 20, "pizarroso" to 10)
        assertEquals(calificador.calificar(conteo).resultado, c2.calificar(conteo).resultado)
    }

    @Test
    fun `avisa cuando se cuentan menos de 97 o mas de 103 granos`() {
        val pocos = calificador.calificar(mapOf("bien_fermentado" to 80))
        assertTrue(pocos.avisos.any { it.contains("80 granos") })

        val muchos = calificador.calificar(mapOf("bien_fermentado" to 110))
        assertTrue(muchos.avisos.any { it.contains("110 granos") })

        val justos = calificador.calificar(
            mapOf("bien_fermentado" to 80, "violeta" to 20),
        )
        assertTrue(justos.avisos.isEmpty())
    }

    @Test
    fun `el formato de numeros no depende de la configuracion regional`() {
        // Si se usara String.format, en un teléfono con coma decimal el texto
        // saldría distinto al de Python y los tests compartidos fallarían.
        val r = RequisitoNorma("fermentado_total", "min", 75.0)
        assertEquals(
            "fermentado total = 70.0 % (requiere ≥ 75 %)",
            r.explicarFalla(70.0),
        )
    }
}
